### PR TITLE
Use ES modules + build and optimize Minimap using Parcel 

### PR DESCRIPTION
**Pull request by aminya**
- This PR Uses ES modules to import/export functions and classes 

- This prevents the need for using esoteric include/mixin methods for combining classes. The time spent during runtime for combining is optimized away using Parcel. 

	In this PR, I only converted Main and PluginManagement into a module. But in the future PRs, I will do it for the other ones, so hopefully will not need mixin hackery.

- It fixes the race conditions that existed before this. Main was being lazy required in a couple of places even when the class was not yet initialized. Because of this, a hack had been implemented to remove Node's require cache, so that every call to `require()` being considered new!!

- By not using classes and exporting everything, we allow the compilers (Chrome, V8, Parcel) to inline the functions and values. Because in JavaScript all the methods/props of a class are public, the engines cannot inline the calls to these functions, and so a lot of performance is killed. However, by using non-exported functions that are local to the module, we allow the compilers to inline the calls to them, and so we will have a much faster runtime.

   Here I only did this for PluginManagement and Main, but in the future PRs, we can allow the compilers to inline all the functions that do not need direct access to the states of the class. If we do this for MinimapElement for example, we can see some performance benefits in the actual Minimap rendering.

- This allows using asynchronous importing instead of synchronous `require` for lazy loading. For example, the plugin generator is now imported asynchronously which does not block the main thread and so it decreases the load time.

### CODE REVIEW

**Comment by UziTech** [comment]
Before we create a build step can we get some benchmarks to show if it is actually faster? If adding a build step saves less than 100ms I don't think it is worth it.

**Comment by aminya** [comment]
@UziTech 

I, as the main maintainer of the minimap package, am totally fine with firing up `npm run dev` in a terminal before doing my development. That has zero effect on my workflow.

Minimap is at a degree of stability that even squeezing 1 more percent out of the code is valuable. I don't expect the code for Minimap to change much. Making minimap even a few percent faster is so important that I sometimes consider converting the package to AssemblyScript and Web assembly.

**Comment by UziTech** [comment]
> I, as the main maintainer of the minimap package,

The problem is that you are not the only one working on it.

Build steps make a burden for people to help create PRs and work on things that you or I don't have time to work on.

A speed up that no one notices would not help anyone.

**Comment by aminya** [comment]
> Build steps make a burden for people to help create PRs and work on things that you or I don't have time to work on.

I don't think that is very true. The surveys show that TypeScript (being compiled) is one of the most loved/growing languages.
https://insights.stackoverflow.com/survey/2019#most-loved-dreaded-and-wanted

Build steps are used to generate bug-free and performant binaries/executables. This is what the "end-users" appreciate. I don't consider `npm run dev` a burden really for me or other few developers that might show up here. 

Arguing that JavaScript is better than TypeScript because it defers everything to the runtime will not reach any meaningful result.

The problems of JavaScript have made big companies such as Microsoft, Google, etc try to build TypeScript, Dart, Flow, Deno, etc for the hope that they can finally kill the unsafe and sluggish languages such as JavaScript.

> A speed up that no one notices would not help anyone.

The end-users will notice it.  See the issues for a glimpse of the complaints about Minimap.

https://github.com/atom-minimap/minimap/issues?q=is%3Aissue+sort%3Aupdated-desc+is%3Aclosed+label%3Aperformance+

**Comment by UziTech** [comment]
> The end-users will notice it.

All I am asking for is proof of this statement with a benchmark test. 

> Arguing that JavaScript is better than TypeScript

I never said that. Also Typescript isn't necessarily faster than javascript.


Is the goal speed or using TypeScript?

**Comment by aminya** [comment]
> > The end-users will notice it.
> 
> All I am asking for is proof of this statement with a benchmark test.
> 

Currently, I am fixing the weird race conditions that had happened because of all this hackery made into Minimap for no reason. 

> > Arguing that JavaScript is better than TypeScript
> 
> Also Typescript isn't necessarily faster than javascript.
> 

Not inherently, but it provides so many ways to detect bugs that result in performance issues and other types of bugs.

> Is the goal speed or using TypeScript?

Your concern was that build-steps are not liked by the developers. I provided you some evidence that it is not the case for everyone, and actually, developers are leaving interpreted languages in favor of compiled ones.

**Comment by UziTech** [comment]
> Your concern was that build-steps are not liked by the developers.

That was not my concern. It is a fact that, if everything else is the same, having a build step requires more work than not having a build step. I understand the benefits of having a build step and using compiled languages. I just want to be sure that in this specific instance the benefits outweigh the costs.

You have not provided any evidence that everything else is not the same (or worse). That is my concern.

**Comment by UziTech** [comment]
Also if the goal is typescript, that can also be accomplished without a build step. Parcel and other bundlers are great for the web because of their reduced time of downloading the javascript but that metric does not matter at all for atom packages since they are always local. The only speed improvement can be load speed or runtime speed which are not guaranteed by bundling javascript.

File: lib/main.js [hunk 1]
```diff
@@ -25,383 +18,349 @@ let Emitter, CompositeDisposable, Minimap, MinimapElement, MinimapPluginGenerato
...
 
-  /**
+/**
    * Opens the plugin generation view.
    *
    * @param  {string} template the name of the template to use
    */
-  generatePlugin (template) {
-    if (!MinimapPluginGeneratorElement) {
-      MinimapPluginGeneratorElement = require('./minimap-plugin-generator-element')
```
Comment (UziTech) [inline lib/main.js:196]: It is almost certainly faster to lazy load `'./minimap-plugin-generator-element'` since most users will never use it.

Comment (aminya) [reply lib/main.js:196]: Yes, thanks. I wanted to go through the code and find these parts that are not used in the default workflow.

The previous code was making literally everything lazy even it was not! This had resulted in a lot of race conditions, and the need for applying hacks for modifying the require cache.

Another module that I found that is not needed is the LegacyAdapter

Comment (aminya) [reply lib/main.js:196]: Fixed in https://github.com/atom-minimap/minimap/pull/736/commits/c7d058dcc60d33f25138ba92ecd90777fd5e99dc

**Comment by UziTech** [comment]
It looks like the build step is missed by semantic release.

I still don't understand what the benefits of parcel here. The rest of the changes could work just fine without a build step and it is just as fast according to my tests.

**Comment by aminya** [comment]
> It looks like the build step is missed by semantic release.

Should be fixed now. 

> I still don't understand what the benefits of parcel here. The rest of the changes could work just fine without a build step and it is just as fast according to my tests.

I added some information to the PR's description. Not using Parcel/Terser makes these changes only syntax sugar. Using Babel to convert import/export to require is like doing nothing. But using Parcel/Terser to inline the functions/import/exports results in the benefits of these changes.

**Comment by UziTech** [comment]
But there is no speed increase so all of these changes are just syntax sugar.

**Comment by UziTech** [comment]
This is actually slower to load than before so again, what is the benefit of these changes?

**Comment by aminya** [comment]
@UziTech

> But there is no speed increase so all of these changes are just syntax sugar.

There is. Inlining, branch removal, etc is very important and affects the performance. But remember that this is an incremental process. I have to go through each file and optimize the things.  

In this PR I just did minimal changes. Look at #737 and #738 for some of the things we can do now that "actually" improve the runtime performance.

> This is actually slower to load than before so again, what is the benefit of these changes?

I am more concerned with the runtime performance. The load time is not much important. It is only 20ms on my machine.

**Comment by UziTech** [comment]
I still think in the end there will not be any speed improvement. Minimizing javascript does not make runtime faster.

**Comment by UziTech** [comment]
You keep changing what you say that you are optimizing. If you want to do all this work for nothing I'm not going to stop you, but usually knowing what your goal is will save you time in the long run. It seems to me that you are just trying to make the code conform to some arbitrary standard that you have in your head. As long as you are willing to put in the work to fix any bugs this might bring I am ok with it.

**Comment by aminya** [comment]
> You keep changing what you say that you are optimizing.

**Comment by UziTech** [comment]
> "proven patterns"

For web development maybe, not necessarily Atom package development. That is an important difference.

I agree that some things can be done to improve performance. I don't take for granted that those improvements outweigh the cost of a build step. Maybe some of these changes will make a difference but if you are not writing tests to make sure performance is *perceivably* better, and not going to be regressed in the future, than it is just wasting time.

**Pull request merged**